addons/l10n_jo_edi/models/account_move.py

```diff
@@ -0,0 +1,258 @@
+                    "name": invoice._l10n_jo_edi_get_xml_attachment_name(),
+                    "datas": base64.b64encode(invoice.env['account.edi.xml.ubl_21.jo']._export_invoice(invoice)),
+                }
+            )
+            invoice.invalidate_recordset(
+                fnames=[
+                    "l10n_jo_edi_xml_attachment_id",
+                    "l10n_jo_edi_xml_attachment_file",
+                ]
+            )
+
+            try:
+                invoice.submit_to_jofotara()
+                invoice.l10n_jo_edi_error = False
+                invoice.with_context(no_new_invoice=True).message_post(
+                    body=_("E-invoice (JoFotara) submitted successfully."),
+                    attachment_ids=invoice_xml.ids,
+                )
+            except (ValidationError, UserError) as error:
+                invoice.l10n_jo_edi_error = error
+                invoice.with_context(no_new_invoice=True).message_post(
+                    body=_("E-invoice (JoFotara) submission failed. %s", error),
+                    attachment_ids=invoice_xml.ids,
+                )
+                return error
+
+    @api.model
+    def create(self, vals):
+        records = super().create(vals)
+        for record in records:
+            if record.move_type in ['out_invoice', 'out_refund'] and record.company_id.country_id.code == 'JO':
+                record.l10n_jo_edi_uuid = uuid.uuid4()
+        return records
+
+
+class AccountMoveLine(models.Model):
+    _inherit = 'account.move.line'
+
+    def _get_conversion_rate(self):
+        return self.currency_rate or 1
+
+    def _get_discount_amount_jod(self):
+        self.ensure_one()
+        discount_amount = 0
+        if self.discount != 0:
+            discount_amount = self._get_tax_exclusive_amount_jod() * (self.discount / 100)
+        return float_round(discount_amount, 9)
+
+    def _get_tax_exclusive_amount_jod(self):
+        self.ensure_one()
+        return float_round(self._get_unit_price_jod() * self.quantity, 9)
+
+    def _get_tax_inclusive_amount_jod(self):
+        self.ensure_one()
+        return float_round(self._get_tax_exclusive_amount_jod() + self._get_special_tax_amount_jod() + self._get_general_tax_amount_jod() - self._get_discount_amount_jod(), 9)
+
+    def _get_special_tax_amount_jod(self):
+        self.ensure_one()
+        for tax in self.tax_ids:
+            if tax.get_jo_tax_type() == 'special':
+                return tax.amount / self._get_conversion_rate()
+        return 0
+
+    def _get_general_tax_amount_jod(self):
+        self.ensure_one()
+        for tax in self.tax_ids:
+            if tax.get_jo_tax_type() == 'general':
+                return float_round((self._get_tax_exclusive_amount_jod() - self._get_discount_amount_jod() + self._get_special_tax_amount_jod()) * (tax.amount / 100), 9)
+        return 0
+
+    def _get_unit_price_jod(self):
+        return float_round(self.price_unit / self._get_conversion_rate(), 9)
```

Comment:
That part is really unclear. It's just helpers about the xml. Just put them in your ubl model then.